`FixedRectangularBinning` symbolization and `Diversity` probabilities

kahaaga opened the pull request:
# What's new in this PR? 

This PR implements the diversity entropy (Wang et al., 2020), and its underlying probability estimator.

## Changes/additions:

- `FixedRectangularBinning` symbolization scheme, which is required for the new method (fixes #118).
- `alphabet_length` implemented for both `RectangularBinning` and `FixedRectangularBinning` (fixes #103).
- Added `Diversity` probability estimator, with analytical tests from the paper.
- Documentation entries for the new `FixedRectangularBinning` and `Diversity`.
- Modified docstring for `RectangularBinning` to highlight how it is different from `FixedRectangularBinning`.

## References
Wang, X., Si, S., & Li, Y. (2020). Multiscale diversity entropy: A novel dynamical measure for fault diagnosis of rotating machinery. IEEE Transactions on Industrial Informatics, 17(8), 5419-5429.

Datseris commented:
(tag me explicitly when you think this is final and ready to review)

kahaaga commented:
> (tag me explicitly when you think this is final and ready to review)

@Datseris If the tests pass and the documentation looks good after the latest commit, this is ready for reviewing.

Datseris commented on src/probabilities_estimators/diversity/diversity.jl:
```diff
@@ -0,0 +1,75 @@
+    return ds, binning
+end
+
+function probabilities(x::AbstractVector{T}, est::Diversity) where T <: Real
+    ds, binning = similarities_and_binning(x, est)
+    bin_estimator = ValueHistogram(binning)
+
+    return probabilities(ds, bin_estimator.binning)
+end
```
Furthermore I thought we completely dissallowed use `probabilities` with anything else besides `<: ProbaiblitiesEstimator`. So how does this call using a `binning `works?

kahaaga replied:
> Furthermore I thought we completely dissallowed use probabilities with anything else besides <: ProbaiblitiesEstimator. 

I can't remember anymore where the discussion ended up. There were probably some good arguments for/against in a  PR/issue somewhere.

> So how does this call using a binning works?

It just does
```julia
function probabilities(x::Vector_or_Dataset, binning::Floating_or_Fixed_RectBinning)
    fasthist!(x, binning)[1]
end
```

which is convenient, because `fasthist` isn't exported and writing `probabilities(x, ValueHistogram(RectangularBinning())` is verbose.

If we don't want to have these convenience methods, it's straight-forward to just remove them and let 

```julia
function probabilities(x::AbstractVector{T}, est::Diversity) where T <: Real
    ds, binning = similarities_and_binning(x, est)
    return fasthist!(ds, binning)[1]
end
```

 and the same for `probabilities_and_outcomes`.

Datseris replied:
ValueHistogram(RectangularBinning()) is verbose. Yes, but we have allowed `ValueHistogram(ε)` to initialize the binning internally. I think we should stick with not allowing `probabilities` to take anything other than an estimator.

Datseris commented on src/probabilities_estimators/diversity/diversity.jl:
```diff
@@ -0,0 +1,75 @@
+
+function probabilities_and_outcomes(x::AbstractVector{T}, est::Diversity) where T <: Real
+    ds, binning = similarities_and_binning(x, est)
+    bin_estimator = ValueHistogram(binning)
+    return probabilities_and_outcomes(ds, bin_estimator.binning)
```
same story as above

Datseris commented on src/probabilities_estimators/histograms/rectangular_binning.jl:
```diff
@@ -15,35 +18,113 @@ Binning instructions are deduced from the type of `ϵ` as follows:
     intervals that cover all data.
 4. `ϵ::Vector{Float64}` divides the i-th coordinate axis into intervals of fixed size
     `ϵ[i]`, starting from the axis minima until the data is completely covered by boxes.
+
+!!! note "Be aware when mixing methods"
+    If the range of the data along some dimension is `[r1, r2]`, then
+    `RectangularBinning(N)` results in a bin edge length of
+    `nextfloat(r2 - r1) / N)`. `nextfloat` is used to ensure data are completely
+    covered. Keep this in mind if you're switching between the `ϵ::Int` and `ϵ::Float64`
+    ways of constructing grids.
```
I don't understand what this comment is trying to achieve. I think the goal is to educate on how the binning works if `n` is Int. But then, why not just state exactly what we do in the 1. bullet point, which would save us 3 lines of documentation? Also, I thought we take the nextfloat twice in the source code.

kahaaga replied:
If I remember correctly, it originated back when using an integer number of bins gave a different number of bins that using exact bin sizes. But this is fixed now. An honestly, upon reading this comment now, it doesn't make much sense. I think we can just remove this comment.

Datseris commented on src/probabilities_estimators/histograms/rectangular_binning.jl:
```diff
@@ -15,35 +18,113 @@ Binning instructions are deduced from the type of `ϵ` as follows:
+
 """
-    RectangularBinMapping <: Encoding
+    FixedRectangularBinning <: AbstractBinning
```
```suggestion
    FixedRectangularBinning <: Encoding
```
This is the information that is actually useful for a user.

Datseris commented on src/probabilities_estimators/histograms/rectangular_binning.jl:
```diff
@@ -15,35 +18,113 @@ Binning instructions are deduced from the type of `ϵ` as follows:
+
+Binning instructions are deduced from the types of `ϵmin`/`emax` as follows:
+
+1. `E::Float64` sets the grid range along all dimensions to to `[ϵmin, ϵmax]`.
+2. `E::NTuple{D, Float64}` sets ranges along each dimension individually, i.e.
+    `[ϵmin[i], ϵmax[i]]` is the range along the `i`-th dimension.
```
This is incorrect syntax. Notice that `E` is already the type of `emin` So `E` will never be `Float64`, as `E` is a data type. What you want to write is `E <: Real` and `E <: Tuple`.

Datseris commented on src/probabilities_estimators/histograms/rectangular_binning.jl:
```diff
@@ -15,35 +18,113 @@ Binning instructions are deduced from the type of `ϵ` as follows:
+    function FixedRectangularBinning(ϵmin::E1, ϵmax::E2, N::Int) where {E1 <: ValidFixedBinInputs, E2 <: ValidFixedBinInputs}
+        f_ϵmin = float.(ϵmin)
+        f_ϵmax = float.(ϵmax)
+        return new{typeof(f_ϵmin)}(float.(f_ϵmin), float.(f_ϵmax), N::Int)
```
you take `float` twice here, `f_emin` is already floated.

Datseris commented on src/probabilities_estimators/histograms/rectangular_binning.jl:
```diff
@@ -81,12 +162,81 @@ function RectangularBinMapping(x::AbstractVector{<:Real}, b::RectangularBinning;
+
+# When the grid is fixed by the user, we can always deduce the total number of bins,
+# even just from the binning itself - symbolization info not needed.
+const FRB = FixedRectangularBinning
+total_outcomes(::AbstractVector, b::FRB) = b.N
```
Then you need to define `total_outcomes(b::FRB)` as well.

Datseris reviewed (changes requested):
What is `RectangularBinMapping`? Where does it come from? Does it replace the `RectangularBinEncoder` I have created? If so, why is the git diff like this? It never shows the `RectangularBinEncoder` being replcaed by the `Mapping`.

p.s.: The fact that this new `FixedRectangularBinning` was written in the same file as the normal binning _and was written in the lines before the normal binning_ made the git diff unecessarily difficult to review. If you check, it seems like most of the `RectangularBinning` code is overwritten/changed in this PR.

kahaaga commented:
> What is RectangularBinMapping? Where does it come from? Does it replace the RectangularBinEncoder I have created? If so, why is the git diff like this? It never shows the RectangularBinEncoder being replcaed by the Mapping.

The `RectangularBinMapping` is exactly `RectangularBinEncoder`, just renamed, so that all encoding schemes have a common naming convention (like `GaussianMapping` and `OrdinalMapping`). Nothing else should have changed.

> p.s.: The fact that this new FixedRectangularBinning was written in the same file as the normal binning and was written in the lines before the normal binning made the git diff unecessarily difficult to review. If you check, it seems like most of the RectangularBinning code is overwritten/changed in this PR.

The git diff became weird when I merged `main` into the `diversity_probs` branch. I guess it was because of what you said about the new lines added before the existing ones?

Datseris commented on test/utils/encoding.jl:
```diff
@@ -1,4 +1,5 @@
-using Entropies.DelayEmbeddings
+using DelayEmbeddings: genembed, Dataset
```
We should make `Entropies` reexport `DelayEmbeddings` I guess

Datseris commented:
> The RectangularBinMapping is exactly RectangularBinEncoder, just renamed, so that all encoding schemes have a common naming convention (like GaussianMapping and OrdinalMapping). 

But these aren't intuitive names though. OrdinalPatterns is more intuitive than OrdinalMapping. I would argue that once again we violate the "one word per concept principle". Why are we using `Mapping` here? Aren't these `Encoding`s? We decided on the Encoding word, we might as well use it. Hence, `RectangularBinEncoding`, `OrdinalPatternEncoding`, etc.

kahaaga commented:
Ok, I think I addressed all your comments now, @Datseris. The CI docs still fail to run, though, and it seems to be related to the use of ChaosTools in the docs.

Datseris commented:
ill take care of the docs in a different PR
